### PR TITLE
add gradient llms and example

### PR DESCRIPTION
**Pull request by lilwilsond2**
# Description

Adds [Gradient](https://gradient.ai/) integration

## Type of Change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
  -- test would need access token and workspace, wasn't how to approach
- [X] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
  -- no new failures, at least

### CODE REVIEW

File: llama_index/llms/gradient.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+class _BaseGradientLLM(CustomLLM):
+    _gradient = PrivateAttr()
+    _model = PrivateAttr()
+
+    # Config
+    max_tokens: Optional[int] = Field(
+        description="The number of tokens to generate.",
+        gt=0,
+        lt=512,
```
Comment (Disiok) [inline llama_index/llms/gradient.py:24]: curious: why enforce a limit of 512 output tokens?

File: llama_index/llms/gradient.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        return CompletionResponse(
+            text=self._model.complete(
+                query=prompt,
+                max_generated_token_count=self.max_tokens,
+                **kwargs,
+            ).generated_output
+        )
+
+    @override
+    def stream_complete(
```
Comment (Disiok) [inline llama_index/llms/gradient.py:83]: streaming API to come in the future?

File: llama_index/llms/gradient.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        prompt: str,
+        **kwargs: Any,
+    ) -> CompletionResponseGen:
+        raise NotImplementedError()
+
+    @property
+    @override
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=1024,
```
Comment (Disiok) [inline llama_index/llms/gradient.py:94]: do all models gradient support have context window of only 1k?

File: llama_index/llms/gradient.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            num_output=self.max_tokens or 20,
```
Comment (Disiok) [inline llama_index/llms/gradient.py:95]: curious: why set to 20?

**Comment by Disiok** [comment]
Thanks for the contribution! Left some comments/questions.

File: llama_index/llms/gradient.py (hunk 1 above)
Comment (lilwilsond2) [reply llama_index/llms/gradient.py:24]: our current api has the same limit

clipped from our public openapi spec
```
"maxGeneratedTokenCount": {
  "type": "integer",
  "nullable": true,
  "minimum": 0,
  "exclusiveMinimum": true,
  "maximum": 512,
  "exclusiveMaximum": true,
  "description": "The maximum number of tokens to generate."
}
```

File: llama_index/llms/gradient.py (hunk 2 above)
Comment (lilwilsond2) [reply llama_index/llms/gradient.py:83]: yeah, our platform doesn't currently have a streaming option

File: llama_index/llms/gradient.py (hunk 3 above)
Comment (lilwilsond2) [reply llama_index/llms/gradient.py:94]: our llama2 and nous-hermes2 offerings have the same 1024 limit for the time being
the bloom-560m offering might have slightly larger, but that's really only good for proof of concept anyway

File: llama_index/llms/gradient.py (hunk 4 above)
Comment (lilwilsond2) [reply llama_index/llms/gradient.py:95]: mostly just for documentation, if you don't explicitly set the output to higher on our platform you'll get 20

**Comment by Disiok** [comment]
Thanks for the clarification on my questions! Would love to get this in. Looks like there's just some minor linting and testing issues left.

**Comment by jerryjliu** [comment]
@lilwilsond2 sorry for the delay on this! seems like there's a slight merge conflict, after that it's ready to merge

**Pull request merged**